drivers: timer: infineon pdl lp_timer

billwatersiii opened the pull request:
Add PDL-based low-power timer for the E84 board

billwatersiii committed:
drivers: timer: infineon pdl lp_timer

Add PDL-based low-power timer for the E84 board

Signed-off-by: Bill Waters <bill.waters@infineon.com>

sreeramIfx commented on drivers/timer/ifx_cat1_lp_timer_pdl.c:
```diff
@@ -0,0 +1,343 @@
+	uint8_t irq_priority;
+};
+
+#define LPTIMER_FREQ (32768u)
```
This should be coming from the DTS file

billwatersiii commented on drivers/timer/ifx_cat1_lp_timer_pdl.c:
```diff
@@ -0,0 +1,343 @@
+	 * case: (2 / 100) * ((150x10^6)/1) = 3x10^6 iterations Compromise: (2 / 100) *
+	 * ((150x10^6)/0xFFFF iterations) = 45 Hz = LFClk_min
+	 */
+	const uint32_t DEFAULT_TIMEOUT = 0xFFFFUL;
```
Done

billwatersiii commented on drivers/timer/ifx_cat1_lp_timer_pdl.c:
```diff
@@ -0,0 +1,343 @@
+	return (uint32_t)ret;
+}
+
+static void lptimer_isr(const struct device *dev)
```
This is ok because it is static

billwatersiii commented on drivers/timer/ifx_cat1_lp_timer_pdl.c:
```diff
@@ -0,0 +1,343 @@
+	}
+}
+
+static int lptimer_init(const struct device *dev)
```
This is ok because it is static

teburd commented on drivers/timer/ifx_cat1_lp_timer_pdl.c:
```diff
@@ -0,0 +1,351 @@
+		.clock_frequency = DT_INST_PROP(n, clock_frequency),                               \
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(n, lptimer_init, NULL, &ifx_cat1_lptimer_data##n,                    \
```
No need to make the timer a device, its a singleton in Zephyr's view. Just a couple of static variables tied to the compilation unit for the hardware timer being used as the kernel timer.

Take a look at how [cortex_m_systick.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/timer/cortex_m_systick.c). It works by  SYS_INIT and a few static variables tied to the compilation unit, along with the appropriate sys_clock functions implemented.

teburd commented on drivers/timer/ifx_cat1_lp_timer_pdl.c:
```diff
@@ -0,0 +1,351 @@
+
+static int lptimer_init(const struct device *dev)
+{
+#define LPTIMER_CTRL (CY_MCWDT_CTR0 | CY_MCWDT_CTR1 | CY_MCWDT_CTR2)
```
The #define in the function body looks a bit odd, I'd shift it out or make it a const value if the only use is in this function.

teburd commented on drivers/timer/ifx_cat1_lp_timer_pdl.c:
```diff
@@ -0,0 +1,351 @@
+	if ((data->isr_instruction & LPTIMER_ISR_CALL_USER_CB_MASK) != 0) {
+		/* announce the elapsed time in ms */
+		uint32_t lptimer_value = Cy_MCWDT_GetCount(config->reg_addr, CY_MCWDT_COUNTER2);
+		int32_t delta_ticks = ((uint64_t)(lptimer_value - last_lptimer_value) *
```
This math is also not quite right. It should be something like...

```c
uint32_t current_cycle_count = sys_read32(SOME_TIMER_COUNT_REG);
uint32_t delta_ticks = (current_cycle_count - announced_cycle_count)/cycles_per_tick;

announced_cycle_count = delta_ticks*cycles_per_tick;

sys_clock_announce(....);

```

teburd reviewed (changes requested):
I think there was some fundamental misunderstanding of how timers work in Zephyr when this was written, and I see it was mostly copied from the existing cat1-lp-timer driver and updated to the new HAL API if I understand this right. So likely both need fixing to correctly work.

This is maybe one of those few occasions where only the base address of the register block for the IP is needed and the few registers (typically 2-3..) are kept in C and Zephyr #define's rather than using the HAL.

billwatersiii commented on dts/bindings/timer/infineon,cat1-lp-timer-pdl.yaml:
```diff
@@ -0,0 +1,20 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+description: Infineon CAT1 lop power timer
```
Done

billwatersiii commented on drivers/timer/ifx_cat1_lp_timer_pdl.c:
```diff
@@ -0,0 +1,351 @@
+	}
+
+	if ((data->isr_instruction & LPTIMER_ISR_CALL_USER_CB_MASK) != 0) {
+		/* announce the elapsed time in ms */
```
Corrected.

billwatersiii commented on drivers/timer/ifx_cat1_lp_timer_pdl.c:
```diff
@@ -0,0 +1,351 @@
+	lp_time = Cy_MCWDT_GetCount(config->reg_addr, CY_MCWDT_COUNTER2);
+
+	/* Convert lptim count in a nb of hw cycles with precision */
+	ret = ((uint64_t)lp_time * sys_clock_hw_cycles_per_sec()) / config->clock_frequency;
```
Corrected

billwatersiii commented on dts/bindings/timer/infineon,cat1-lp-timer-pdl.yaml:
```diff
@@ -0,0 +1,20 @@
+
+description: Infineon CAT1 lop power timer
+
+compatible: "infineon,cat1-lp-timer-pdl"
```
This applies to a number of our drivers.  We have a task already captured to make this update and create a pull request with the update for all of them at once.  So I will let this be taken care of in that other pull request.

jsbatch commented on drivers/timer/infineon_lp_timer_pdl.c:
```diff
@@ -0,0 +1,369 @@
+							  .c0c1Cascade = true,
+							  .c1c2Cascade = false};
+
+static uint32_t last_lptimer_value;
```
last_lptimer_value is never initialized.

mcatee-infineon replied:
This should be fixed now.

jsbatch commented on drivers/timer/infineon_lp_timer_pdl.c:
```diff
@@ -0,0 +1,369 @@
+	} else {
+		Cy_MCWDT_Disable(reg_addr, LPTIMER_COUNTERS, LPTIMER_RESET_TIME_US);
+		Cy_MCWDT_DeInit(reg_addr);
+	}
```
Failure case doesn't result in an error return.  Also, the IRQ is enabled, even if the init function fails.

teburd reviewed (changes requested):
Driver works, but needs some Kconfig finagling today to get it enabled for pse84

Take a look at how other boards/socs do this.

Attached my patch to get it working, along with manually de-selecting cortex-m-systick which I should not need to do, ideally this is a SOC Kconfig Option with the default being LP Timer.

Test results once I ran this on pse84

```sh
*** Booting Zephyr OS build v4.3.0-2098-g816fd8a57d10 ***
Running TESTSUITE timer_jitter_drift
===================================================================
START - test_jitter_drift_timer_period
periodic timer behavior test using built-in restart mechanism
collecting time samples for approx 10 seconds
periodic timer samples gathered, calculating statistics
timer clock rate 32768, kernel tick rate 10000
expected period drift: -84.472656 us
period duration statistics for 10000 samples (0 rollovers):
  expected: 1000 us,            32.768000 cycles
  min:      885.009766 us,      29 cycles
  max:      946.044922 us,      31 cycles
  mean:     915.527344 us,      30.000000 cycles
  variance: 0.186265 us,        0.000200 cycles
  stddev:   0.431584 us,        0.014142 cycles
timer start cycle 569, end cycle 300569,
total time 9155273.437500 us, expected time 10000000.000000 us,
expected time drift -844726.562500 us, difference 0.000000 us
RECORD: {"testcase":"jitter_drift_timer", "mechanism":"builtin", "stats_count":10000, "rollovers":0, "mean_us":915.527344}
 PASS - test_jitter_drift_timer_period in 9.297 seconds
===================================================================
START - test_jitter_drift_timer_startdelay
periodic timer behavior test using explicit start with delay
collecting time samples for approx 10 seconds
periodic timer samples gathered, calculating statistics
timer clock rate 32768, kernel tick rate 10000
expected period drift: -84.472656 us
period duration statistics for 10000 samples (0 rollovers):
  expected: 1000 us,            32.768000 cycles
  min:      885.009766 us,      29 cycles
  max:      946.044922 us,      31 cycles
  mean:     915.527344 us,      30.000000 cycles
  variance: 0.186265 us,        0.000200 cycles
  stddev:   0.431584 us,        0.014142 cycles
timer start cycle 305666, end cycle 605666,
total time 9155273.437500 us, expected time 10000000.000000 us,
expected time drift -844726.562500 us, difference 0.000000 us
RECORD: {"testcase":"jitter_drift_timer", "mechanism":"startdelay", "stats_count":10000, "rollovers":0, "mean_us":915.527}
 PASS - test_jitter_drift_timer_startdelay in 9.297 seconds
===================================================================
TESTSUITE timer_jitter_drift succeeded
Running TESTSUITE timer_tick_train
===================================================================
START - test_one_tick_timer_train
Initializing 4 Timers, Tick Rate 10000Hz, Expecting 25000 callbacks in 11000 ms
Starting Timers with Skews
Faking busy work, remaining timers is 4, timer callbacks 0
Faking busy work, remaining timers is 4, timer callbacks 697
Faking busy work, remaining timers is 4, timer callbacks 1380
Faking busy work, remaining timers is 4, timer callbacks 2063
Faking busy work, remaining timers is 4, timer callbacks 2746
Faking busy work, remaining timers is 4, timer callbacks 3429
Faking busy work, remaining timers is 4, timer callbacks 4111
Faking busy work, remaining timers is 4, timer callbacks 4794
Faking busy work, remaining timers is 4, timer callbacks 5477
Faking busy work, remaining timers is 4, timer callbacks 6160
Faking busy work, remaining timers is 4, timer callbacks 6843
Faking busy work, remaining timers is 4, timer callbacks 7526
Faking busy work, remaining timers is 4, timer callbacks 8209
Faking busy work, remaining timers is 4, timer callbacks 8892
Faking busy work, remaining timers is 4, timer callbacks 9574
Faking busy work, remaining timers is 4, timer callbacks 10257
Faking busy work, remaining timers is 4, timer callbacks 10940
Faking busy work, remaining timers is 4, timer callbacks 11623
Faking busy work, remaining timers is 4, timer callbacks 12306
Faking busy work, remaining timers is 4, timer callbacks 12989
Faking busy work, remaining timers is 4, timer callbacks 13672
Faking busy work, remaining timers is 4, timer callbacks 14355
Faking busy work, remaining timers is 4, timer callbacks 15037
Faking busy work, remaining timers is 4, timer callbacks 15720
Faking busy work, remaining timers is 4, timer callbacks 16403
Faking busy work, remaining timers is 4, timer callbacks 17086
Faking busy work, remaining timers is 4, timer callbacks 17769
Faking busy work, remaining timers is 4, timer callbacks 18452
Faking busy work, remaining timers is 4, timer callbacks 19135
Faking busy work, remaining timers is 4, timer callbacks 19818
Faking busy work, remaining timers is 4, timer callbacks 20500
Faking busy work, remaining timers is 4, timer callbacks 21183
Faking busy work, remaining timers is 4, timer callbacks 21866
Faking busy work, remaining timers is 4, timer callbacks 22549
Faking busy work, remaining timers is 4, timer callbacks 23232
Faking busy work, remaining timers is 4, timer callbacks 23915
Faking busy work, remaining timers is 4, timer callbacks 24598
One Tick Timer Train Done, took 10112 ms, busy loop ran 37 times
    Perfect delta 13 cycles or 400 us
Timer 0 max delta 13 cycles or 397 us, 0 late callbacks (0.0%)
RECORD: {"testcase":"one_tick_timer_train", "timer":0, "max_delta_cycles":13, "max_delta_us":397, "late_callbacks":0, "pe}
Timer 1 max delta 13 cycles or 397 us, 0 late callbacks (0.0%)
RECORD: {"testcase":"one_tick_timer_train", "timer":1, "max_delta_cycles":13, "max_delta_us":397, "late_callbacks":0, "pe}
Timer 2 max delta 13 cycles or 397 us, 0 late callbacks (0.0%)
RECORD: {"testcase":"one_tick_timer_train", "timer":2, "max_delta_cycles":13, "max_delta_us":397, "late_callbacks":0, "pe}
Timer 3 max delta 13 cycles or 397 us, 0 late callbacks (0.0%)
RECORD: {"testcase":"one_tick_timer_train", "timer":3, "max_delta_cycles":13, "max_delta_us":397, "late_callbacks":0, "pe}
 PASS - test_one_tick_timer_train in 9.399 seconds
===================================================================
TESTSUITE timer_tick_train succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [timer_jitter_drift]: pass = 2, fail = 0, skip = 0, total = 2 duration = 18.594 seconds
 - PASS - [timer_jitter_drift.test_jitter_drift_timer_period] duration = 9.297 seconds
 - PASS - [timer_jitter_drift.test_jitter_drift_timer_startdelay] duration = 9.297 seconds

SUITE PASS - 100.00% [timer_tick_train]: pass = 1, fail = 0, skip = 0, total = 1 duration = 9.399 seconds
 - PASS - [timer_tick_train.test_one_tick_timer_train] duration = 9.399 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Patch needed (this isn't correct, its hacked up but at least gets you a working test)

```diff
diff --git a/drivers/timer/Kconfig.infineon_lp b/drivers/timer/Kconfig.infineon_lp
index 39085a294ac..36600d3c3a1 100644
--- a/drivers/timer/Kconfig.infineon_lp
+++ b/drivers/timer/Kconfig.infineon_lp
@@ -1,4 +1,4 @@
-# Infineon CAT1 LPTIMER configuration options
+# Infineon LPTIMER configuration options
 
 # Copyright (c) 2025 Cypress Semiconductor Corporation (an Infineon company) or
 # an affiliate of Cypress Semiconductor Corporation
@@ -6,11 +6,10 @@
 # SPDX-License-Identifier: Apache-2.0
 
 config INFINEON_LP_TIMER_HAL
-	bool "Infineon CAT1 Low Power Timer driver"
+	bool "Infineon Low Power Timer driver (Legacy HAL Based)"
 	default y
 	depends on DT_HAS_INFINEON_LP_TIMER_ENABLED
 	depends on USE_INFINEON_LEGACY_HAL
-	depends on PM
 	select USE_INFINEON_LPTIMER
 	select TICKLESS_CAPABLE
 	help
@@ -19,11 +18,10 @@ config INFINEON_LP_TIMER_HAL
 	  is based on the Infineon Legacy HAL.
 
 config INFINEON_LP_TIMER_PDL
-	bool "Infineon CAT1 Low Power Timer driver"
+	bool "Infineon Low Power Timer driver (PDL Based)"
 	default y
 	depends on DT_HAS_INFINEON_LP_TIMER_ENABLED
 	depends on !USE_INFINEON_LEGACY_HAL
-	depends on PM
 	select USE_INFINEON_LPTIMER
 	select TICKLESS_CAPABLE
 	help
diff --git a/dts/arm/infineon/edge/pse84/pse84_s.dtsi b/dts/arm/infineon/edge/pse84/pse84_s.dtsi
index 5cd77495ae2..59f09b65f19 100644
--- a/dts/arm/infineon/edge/pse84/pse84_s.dtsi
+++ b/dts/arm/infineon/edge/pse84/pse84_s.dtsi
@@ -418,7 +418,7 @@
 			compatible = "infineon,lp-timer";
 			reg = <0x5240d000 0x40>;
 			interrupts = <55 4>;
-			status = "disabled";
+			status = "okay";
 		};
 
 		mcwdt1: mcwdt@5240d040 {
diff --git a/soc/infineon/cat1b/cyw20829/Kconfig.defconfig b/soc/infineon/cat1b/cyw20829/Kconfig.defconfig
index 0db0ec84fcb..174f2680a0e 100644
--- a/soc/infineon/cat1b/cyw20829/Kconfig.defconfig
+++ b/soc/infineon/cat1b/cyw20829/Kconfig.defconfig
@@ -7,9 +7,6 @@
 
 if SOC_SERIES_CYW20829
 
-config INFINEON_CAT1_LP_TIMER
-	bool
-
 config CORTEX_M_SYSTICK
 	default n if INFINEON_CAT1_LP_TIMER
 
diff --git a/soc/infineon/edge/pse84/Kconfig.defconfig b/soc/infineon/edge/pse84/Kconfig.defconfig
index b4c6e914162..c6f7432af96 100644
--- a/soc/infineon/edge/pse84/Kconfig.defconfig
+++ b/soc/infineon/edge/pse84/Kconfig.defconfig
@@ -19,8 +19,7 @@ config NUM_IRQS
 	default 174 if CPU_CORTEX_M55
 
 config SYS_CLOCK_HW_CYCLES_PER_SEC
-	default $(dt_node_int_prop_int,/cpus/cpu@0,clock-frequency) if CPU_CORTEX_M33
-	default $(dt_node_int_prop_int,/cpus/cpu@1,clock-frequency) if CPU_CORTEX_M55
+	default 32768
 
 config BUILD_OUTPUT_ADJUST_LMA
 	depends on XIP
```

mcatee-infineon committed:
drivers: enable lp_timer default for kit_psc3m5_evk

- Enable lp_timer for kit_psc3m5_evk
- Modify configuration, enabling lp_timer as default before systick

Signed-off-by: McAtee Maxwell <maxwell.mcatee@infineon.com>

mcatee-infineon commented on drivers/timer/Kconfig.infineon_lp:
```diff
@@ -5,13 +5,27 @@
 
-config INFINEON_CAT1_LP_TIMER
+config INFINEON_LP_TIMER_HAL
 	bool "Infineon CAT1 Low Power Timer driver"
```
This should be fixed now.

teburd reviewed (approved):
Great job, timer drivers can be quite finicky
